### PR TITLE
Lots of tiny incremental simplifications of `EmitterWriter` internals

### PR DESCRIPTION
**Pull request by oli-obk**
ignore the first commit, it's https://github.com/rust-lang/rust/pull/114088 squashed and rebased, but it's needed to use to use `derive_setters`, as they need a newer `syn` version.

Then this PR starts out with removing many arguments that are almost always defaulted to `None` or `false` and replace them with builder methods that can set these fields in the few cases that want to set them.

After that it's one commit after the other that removes or merges things until everything becomes some very simple trait objects

### CODE REVIEW

File: compiler/rustc_errors/src/emitter.rs [hunk 1]
```diff
@@ -2728,40 +2724,6 @@ impl Style {
...
-    fn flush(&mut self) -> io::Result<()> {
-        match *self {
-            WritableDst::Raw(ref mut w) => w.flush(),
-        }
-    }
-}
-
-impl<'a> Drop for WritableDst<'a> {
-    fn drop(&mut self) {
-        self.flush().unwrap()
```
Comment (cjgillot) [inline compiler/rustc_errors/src/emitter.rs]: Should this flush be called manually somewhere?

**Review by compiler-errors — APPROVED** [review]

**Pull request merged**